dump.go

```diff
@@ -0,0 +1,147 @@
+package lua
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+type dumpState struct {
+	l     *State
+	out   io.Writer
+	order binary.ByteOrder
+	err   error
+}
+
+func (d *dumpState) write(data interface{}) {
+	if d.err == nil {
+		d.err = binary.Write(d.out, d.order, data)
+	}
+}
+
+func (d *dumpState) writeInt(i int) {
+	d.write(int32(i))
+}
+
+func (d *dumpState) writePC(p pc) {
+	d.writeInt(int(p))
+}
+
+func (d *dumpState) writeCode(p *prototype) {
+	d.writeInt(len(p.code))
+	d.write(p.code)
+}
+
+func (d *dumpState) writeByte(b byte) {
+	d.write(b)
+}
+
+func (d *dumpState) writeBool(b bool) {
+	if b {
+		d.writeByte(1)
+	} else {
+		d.writeByte(0)
+	}
+}
+
+func (d *dumpState) writeNumber(f float64) {
+	d.write(f)
+}
+
+func (d *dumpState) writeConstants(p *prototype) {
+	d.writeInt(len(p.constants))
+
+	for _, o := range p.constants {
+		d.writeByte(byte(d.l.valueToType(o)))
+
+		switch o := o.(type) {
+		case nil:
+		case bool:
+			d.writeBool(o)
+		case float64:
+			d.writeNumber(o)
+		case string:
+			d.writeString(o)
+		default:
+			d.l.assert(false)
+		}
+	}
+
+	d.writeInt(len(p.prototypes))
+
+	for _, o := range p.prototypes {
+		d.dumpFunction(&o)
+	}
```

Comment:
Could break this out into `writePrototypes` for consistency with `readPrototypes` in `undump.go`.